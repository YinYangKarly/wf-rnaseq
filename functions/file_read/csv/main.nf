def CSV_READ(samples) {
    read_list = []
    samples.each{ instance ->
        id = "${instance.sample}_${instance.library}_${instance.R1_md5}"
        sample = "${instance.sample}"
        read1 = instance.R1
        read2 = instance.R2 != null ? instance.R2 : ""

        if (read2 != "") {
                read_list << [[id: id, single_end: false, sample: sample], [file(read1, checkIfExists: true), file(read2, checkIfExists: true)]]
            }
        else {
                read_list << [[id : id, single_end: true, sample: sample], [file(read1, checkIfExists: true)]]
            }
    }

    return Channel.fromList(read_list)

}