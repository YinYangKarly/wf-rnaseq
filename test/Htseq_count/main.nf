process HTSEQ_COUNT_COMPARISON {
    
    output:
    path("*.txt"), emit: corrolation
    
    """
    #!/usr/bin/env python
    import csv

    def extract_gene_number(header):
        return header.split('gene')[-1]

    with open('/exports/sascstudent/vperinbanathan/WDL_RNA/output/expression_measures/fragments_per_gene/all_samples.fragments_per_gene', 'r') as WDL, open('/exports/sascstudent/vperinbanathan/nextflow_pipeline_testzone/outputfiles/final_gene_count/report/report_collect_0.csv', 'r') as Nextflow:

        reader = csv.reader(WDL, delimiter="\t")
        header = next(reader)
        sorted_header = sorted(header, key=extract_gene_number)
        sorted_indices = [header.index(col) for col in sorted_header]
        rows = list(reader)
        WDL_lines = sorted(rows, key=lambda row: row[0])
        WDL_line = len(WDL_lines) -1
        
        line_count = len(WDL_lines) * len(WDL_lines[0])
        

        reader = csv.reader(Nextflow, delimiter="\t")
        header = next(reader)
        sorted_header = sorted(header, key=extract_gene_number)
        sorted_indices = [header.index(col) for col in sorted_header]
        rows = list(reader)
        Nextflow_lines = sorted(rows, key=lambda row: row[0])
        Nextflow_line = len(Nextflow_lines) -1
        
        line_corrolation = 0

        while len(WDL_lines) != 0 and len(Nextflow_lines) != 0:
            WDL_instance = WDL_lines[WDL_line]
            Nextflow_instance = Nextflow_lines[Nextflow_line]

            if WDL_instance[0] == Nextflow_instance[0]:
                corrolation = 0
                for WDL_value, Nextflow_value in zip(WDL_instance[1:] , Nextflow_instance[1:]):

                    if float(WDL_value) == float(Nextflow_value):
                        corrolation += 1
                    line_corrolation += corrolation / len(WDL_instance[1:])

                del WDL_lines[-1]
                del Nextflow_lines[-1]
                WDL_line -= 1
                Nextflow_line -= 1

            elif int(WDL_instance[0].strip("ENSG")) > int(Nextflow_instance[0].strip("ENSG")):
                WDL_line -= 1
                del WDL_lines[-1]

            else:
                Nextflow_line -= 1
                del Nextflow_lines[-1]
    output_corrolation = line_corrolation/line_count*100
    with open('corrolation.txt', 'w') as file:
        file.write(str(output_corrolation))
    """
}
