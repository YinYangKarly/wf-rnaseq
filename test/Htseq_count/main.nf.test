nextflow_workflow {
    

    name "Test worfklow with Htseqcount"
    script "../../main.nf"
    workflow "RNA_seq"

    stage {
        symlink "/exports/sascstudent/vperinbanathan"
    }

    test("Corrolation of Collect column counts has to be higher than 99.9%") {

        when{
            params{
                outdir = "/exports/sascstudent/vperinbanathan/nextflow_pipeline_testzone/test/outputfiles"
                sampleConfigFile = "/exports/sascstudent/vperinbanathan/nextflow_pipeline_testzone/inputfiles/testfiles_big_data/samplesheet.csv"
                genome = 'GRCh38'
            }
        }

        then {
            def WDL_file_path = "/exports/sascstudent/vperinbanathan/WDL_RNA/output/expression_measures/fragments_per_gene/all_samples.fragments_per_gene"
            def Nextflow_file_path = "${params.outdir}/final_gene_count/report/report_collect_0.csv"

            assert workflow.success

            //CHECK HTSEQ--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
            //WDL INITIALIZING
            def wdl_file_list = []
            def wdl_file = path(WDL_file_path).eachLine {
                wdl_file_list.add(it.split("\t"))
            }

            def header = []
            wdl_file_list[0].each { item ->
                    header.add(item)
                    }
            def sorted_header = wdl_file_list[0].sort()
            def WDL_lines = []
            def indexes = []
            sorted_header.each { column ->
                indexes.add(header.indexOf(column))
            }

            for (int i = 1; i < wdl_file_list.size(); i++) {
                def line = []
                for (part in indexes) {
                    line.add(wdl_file_list[i][part])
                WDL_lines.add(line)

                }
            }
            def WDL_line = WDL_lines.size() -1

            //NEXTFLOW INITIALIZING
            def nextflow_file_list = []
            def nextflow_file = path(Nextflow_file_path).eachLine {
                nextflow_file_list.add(it.split("\t"))
            }
            header = []
            nextflow_file_list[0].each { item ->
            header.add(item)
            }
            sorted_header = nextflow_file_list[0].sort()
            def Nextflow_lines = []
            indexes = []
            sorted_header.each { column ->
                indexes.add(header.indexOf(column))
            }
            for (int i = 1; i < nextflow_file_list.size(); i++) {
                def line = []
                for (part in indexes) {
                    line.add(nextflow_file_list[i][part])
                Nextflow_lines.add(line)

                }
            }
            def Nextflow_line = Nextflow_lines.size() -1

            def line_corrolation = 0
            def line_count = WDL_lines.size()

            while (0 < Nextflow_lines.size() && 0 < WDL_lines.size()) {
                def WDL_instance = WDL_lines[WDL_line]
                def Nextflow_instance = Nextflow_lines[Nextflow_line]
                if (WDL_instance[0] == Nextflow_instance[0]) {
                    def corrolation = 0
                    for (int i= 1; i < Nextflow_instance.size(); i++) {
                        if (Float.valueOf(WDL_instance[i]) == Float.valueOf(Nextflow_instance[i])) {
                            corrolation += 1
                        }
                    }
                    line_corrolation += corrolation / (WDL_instance.size() -1 )
                    WDL_lines.remove(WDL_line)
                    Nextflow_lines.remove(Nextflow_line)
                    WDL_line -= 1
                    Nextflow_line -= 1
                }

                else if(WDL_instance[0].replaceAll(/[^0-9]/, '').toInteger() > Nextflow_instance[0].replaceAll(/[^0-9]/, '').toInteger()) {
                        WDL_lines.remove(WDL_line)
                        WDL_line -= 1
                }

                else {
                    Nextflow_lines.remove(Nextflow_line)
                    Nextflow_line -= 1
                }
            }
            def output_corrolation = line_corrolation/line_count*100
            assert output_corrolation > 99.9
        
            //CHECK HTSEQ_END--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
        }
    }
}