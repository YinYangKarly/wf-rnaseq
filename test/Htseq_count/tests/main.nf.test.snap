{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    "corrolation.txt:md5,22050f6c629b68080ccb5c5a4fa015d2"
                ],
                "corrolation": [
                    "corrolation.txt:md5,22050f6c629b68080ccb5c5a4fa015d2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-06T12:16:15.784678248"
    }
}